indigo_social/models.py

```diff
@@ -1,8 +1,61 @@
 # -*- coding: utf-8 -*-
+import os
+from io import BytesIO
+import sys
+
 from django.contrib.auth.models import User
+from django.core.files import File
+from django.core.files.uploadedfile import InMemoryUploadedFile
 from django.db import models
-from django.db.models.signals import post_save
+from django.db.models.signals import post_save, pre_save
 from django.dispatch import receiver
+from django.urls import reverse
+from PIL import Image
+
+from urllib.request import urlretrieve, urlcleanup
+
+from allauth.account.signals import user_signed_up
+from allauth.socialaccount.models import SocialLogin, SocialAccount
+from allauth.socialaccount.signals import pre_social_login, social_account_added, social_account_updated
+
+
+def user_profile_photo_path(instance, filename):
+    file_name, file_extension = os.path.splitext(filename)
+    return 'avatars/{}'.format(instance.id)
+
+
+def retrieve_social_profile_photo(user_profile, url):
+    """ Retrieve and save a retrieved user's profile photo from a url
+    """
+    try:
+        filename, _ = urlretrieve(url)
+        user_profile.profile_photo.save(filename, File(open(filename, 'rb')))
+    finally:
+        urlcleanup()
+
+
+def resize_photo(profile_photo, user):
+    """ Resize and crop an uploaded profile photo to meet specifications
+    """
+    uploaded_profile_photo = Image.open(profile_photo)
```

Comment:
FYI, the Django docs say you can use `profile_photo.image`. Haven't tried it :)

Reply:
The image is usually stored in `profile_photo.photo` but it still raises an error because it has not been saved yet

Reply:
I suspect this method could be simpler: it could take an `image` object modify it in place. The caller would then be responsible for writing it back to the file object.

Reply:
Alternatively return a new image object. Basically, you can avoid writing files and do it in memory.

Reply:
I have actually moved it to the form, and have it take an image object only instead without all the other context, then returns a processed image